### PR TITLE
Warn if people mutate children.

### PR DESCRIPTION
**Pull request by jimfb**
Warn if people mutate children.

### CODE REVIEW

**Comment by jimfb** [comment]
I've seen this come up enough that we should just kill it with fire.  cc @spicyj @gaearon

**Comment by spicyj** [comment]
I've never seen this. You mean they're mutating the array that React.createElement makes? Can we just freeze it instead?

Public API so @facebook/react-core.

**Comment by jimfb** [comment]
`createElement` doesn't create an array; it just takes in an array and holds a reference to it, right?  So no, I don't think we can just freeze the array.  Best we can do is warn, I think.

But yeah, they're mutating the props.children array.  It's terrible.

**Comment by sebmarkbage** [comment]
createElement does create an array when you use the var args. We could probably freeze that one. Needs to be tracked first if it is common though.

When people mutate their children what's the common pattern?

Do these gets stored in state with elements? That seems fine(ish). Probably used as a perf optimization by experts.

Or is it data stored in state which gets mutated by render to override it with elements?

I assume a common one is that a wrapper adds children to an existing set.

**Comment by jimfb** [comment]
> I assume a common one is that a wrapper adds children to an existing set.

Yes, it is usually when a wrapper is mutating their props.children array.

I've also seen:
```
for(var i = 0; i < props.children.length; i++) {
   props.children[i] = <div>{props.children[i]}</div>;
}
```

I've also seen cases where the element is created, and elsewhere in the code (internally, UFI) there is logic to push elements into the array.  It was a readability disaster of epic proportions.

**Comment by jimfb** [comment]
> createElement does create an array when you use the var args. We could probably freeze that one.

Also, where is varargs?  Signature of `createElement` is `function(type, config, children)`.

**Comment by keyanzhang** [comment]
https://github.com/facebook/react/blob/master/src/isomorphic/classic/element/ReactElement.js#L187

We could update the signature to `function(type, config, ...children)`.

**Comment by jimfb** [comment]
> ReactElement.js#L187

Ok, yeah, I'm fine with freezing that array.  Either way, we probably want to warn for a release before breaking people's apps, right?

I don't really care, if you guys want to turn this into a breaking change, that's fine with me.  Just let me know how to update the PR to move this forward.

**Comment by jimfb** [comment]
Ping @sebmarkbage, as per team sync.

**Comment by facebook-github-bot-9** [comment]
@jimfb updated the pull request.

**Comment by syranide** [comment]
> createElement doesn't create an array; it just takes in an array and holds a reference to it, right? So no, I don't think we can just freeze the array. Best we can do is warn, I think.

Even if it takes an array I don't see the problem. Contractually you're required to not mutate the array so ownership is effectively transferred to React anyway and it should be free to freeze it. Right?

**Comment by satya164** [comment]
>  createElement does create an array when you use the var args. We could probably freeze that one. Needs to be tracked first if it is common though.

If it's freezed, it'll show nothing in loose mode when trying to mutate, and I think it will be confusing.

**Comment by jimfb** [comment]
> Even if it takes an array I don't see the problem. Contractually you're required to not mutate the array so ownership is effectively transferred to React anyway and it should be free to freeze it. Right?

Passing an object to React.createElement does not transfer ownership to React any more than passing the object to `console.log` transfers ownership to `console`.  It would be confusing/surprising/bad if `console.log` decided to freeze whatever you passed in.  Functions should generally avoid mutating their inputs.  Allowing React.createElement to freeze the array is a form of mutation.

**Comment by syranide** [comment]
> Passing an object to React.createElement does not transfer ownership to React any more than passing the object to console.log transfers ownership to console. It would be confusing/surprising/bad if console.log decided to freeze whatever you passed in. Functions should generally avoid mutating their inputs. Allowing React.createElement to freeze the array is a form of mutation.

@jimfb I have to admit I'm not entirely up on the internals here, if modifying `children` is a danger to React internals (which I assumed, but I guess I'm wrong now that I think about it) then it should reasonably be prevented from mutating.

If not, then we have previously said that children should be treated equally to other props. Such that children only have special treatment for convenience and passing the same data through other props should exhibit identical behavior. I'm not sure of your stance on the API internally nowadays, but previously, passing varargs or an array was considered synonymous. Freezing varargs would contradict that. For all intents and purposes, children does not need to be renderables and a component getting data through children rather than a prop can be a visual/convenience preference. So if we don't freeze props then freezing children is a bit odd, I agree it makes sense in the overwhelming majority of cases, but IMHO it does contradict "children being just another prop" if we still stand by it. tl;dr I feel like this boils down to whether we find it ok to consider vararg/JSX-children to be special or if they should be straight-up forwarded as an array. Also, what about nested children?

**Comment by jimfb** [comment]
@syranide Conceptually, I agree with you.  There are edge cases like someone choosing a different name (`children2`) or mutating nested children.  Ultimately this is a pragmatic attempt at catching the common cases of a particularly nasty antipattern.  It's just a (really good) heuristic.

File: src/isomorphic/classic/element/ReactElement.js [hunk 1]
```diff
@@ -116,6 +116,12 @@ var ReactElement = function(type, key, ref, self, source, owner, props) {
         writable: false,
         value: self,
       });
+      Object.defineProperty(element, '_shadowChildren', {
```
Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: Define this in the non canDefineProperty case too.

Comment (jimfb) [inline src/isomorphic/classic/element/ReactElement.js]: Done.

**Comment by facebook-github-bot-5** [comment]
@jimfb updated the pull request.

**Comment by fab1an** [comment]
I wrote a HOC which changes the element-tree ([react-native-style-tachyons](https://github.com/tachyons-css/react-native-style-tachyons)).

I needed a couple of hours to understand when props, children etc. are allowed to modify or not, how to copy `key`s and/or `children`, since there's not really much documentation for functions like `cloneElement`.

See my question on SO: https://stackoverflow.com/questions/37120956/react-cloneelement-pass-new-children-or-copy-props-children

**Comment by spicyj** [comment]
I answered the questions in your SO post.

**Comment by fab1an** [comment]
@spicyj Great thanks.

**Pull request merged**